File: ECHIRON/src/utils/encouragementEngine.ts
import {
  contextById,
  contextDefinitions,
  encouragementPrinciples,
  principleById,
  type EncouragementPrinciple,
  type EncouragementSignal,
} from '@/data/encouragementLibrary';
import type {
  AppData,
  EncouragementContext,
  EncouragementLength,
  EncouragementPreferences,
  EncouragementRecord,
  EncouragementSource,
  Priority,
  UserProfile,
} from '@/types/models';

export type CompletionSubject = {
  sourceType: EncouragementSource;
  sourceId: string;
  title: string;
  details: string;
  category: string;
  priority: Priority;
  createdAt: string;
};

export type EncouragementDraft = {
  context: EncouragementContext;
  principleId: string;
  principleTitle: string;
  messageId: string;
  heading: string;
  message: string;
};

export type LessonSummary = {
  principleId: string;
  title: string;
  summary: string;
  count: number;
  savedCount: number;
};

export type WeeklyReflection = {
  heading: string;
  message: string;
  completedCount: number;
  highPriorityCount: number;
  focusCount: number;
  dominantContextLabel: string | null;
};

export const defaultEncouragementPreferences: EncouragementPreferences = {
  tone: 'balanced',
  frequency: 'every',
  length: 'brief',
  usePreferredName: true,
  spiritualEnabled: false,
  animationsEnabled: true,
  pinnedPrincipleIds: [],
};

const stopWords = new Set([
  'a',
  'an',
  'and',
  'are',
  'as',
  'at',
  'be',
  'by',
  'for',
  'from',
  'in',
  'is',
  'it',
  'my',
  'of',
  'on',
  'or',
  'the',
  'to',
  'with',
  'your',
]);

export function normalizeEncouragementPreferences(
  preferences?: Partial<EncouragementPreferences> | null,
): EncouragementPreferences {
  return {
    ...defaultEncouragementPreferences,
    ...preferences,
    pinnedPrincipleIds: Array.isArray(preferences?.pinnedPrincipleIds)
      ? preferences.pinnedPrincipleIds
      : [],
  };
}

function normalizeText(value: string) {
  return value.toLowerCase().replace(/[^a-z0-9\s-]/g, ' ');
}

function includesKeyword(text: string, keyword: string) {
  const normalizedKeyword = normalizeText(keyword).trim();
  if (!normalizedKeyword) return false;

  if (normalizedKeyword.includes(' ') || normalizedKeyword.includes('-')) {
    return text.includes(normalizedKeyword);
  }

  const tokens = new Set(text.split(/\s+/).filter(Boolean));
  return tokens.has(normalizedKeyword);
}

function meaningfulWords(value: string) {
  return new Set(
    normalizeText(value)
      .split(/\s+/)
      .filter((word) => word.length >= 4 && !stopWords.has(word)),
  );
}

function classifyContext(
  subject: CompletionSubject,
  profile: UserProfile | null,
): EncouragementContext {
  const categoryText = normalizeText(subject.category);
  const titleText = normalizeText(subject.title);
  const detailsText = normalizeText(subject.details);
  const profileText = normalizeText(
    `${profile?.role ?? ''} ${profile?.goals ?? ''}`,
  );

  let bestContext: EncouragementContext = 'general';
  let bestScore = 0;

  contextDefinitions.forEach((definition) => {
    if (definition.id === 'general') return;

    let score = 0;
    definition.keywords.forEach((keyword) => {
      if (includesKeyword(categoryText, keyword)) score += 5;
      if (includesKeyword(titleText, keyword)) score += 3;
      if (includesKeyword(detailsText, keyword)) score += 1;
      if (includesKeyword(profileText, keyword)) score += 0.5;
    });

    if (score > bestScore) {
      bestScore = score;
      bestContext = definition.id;
    }
  });

  if (subject.sourceType === 'focus') return 'focus';
  return bestContext;
}

function isSameLocalDay(left: string | Date, right: string | Date) {
  return new Date(left).toDateString() === new Date(right).toDateString();
}

function hoursBetween(earlier: string, later: string) {
  return (
    (new Date(later).getTime() - new Date(earlier).getTime()) /
    (1000 * 60 * 60)
  );
}

function isGoalAligned(subject: CompletionSubject, profile: UserProfile | null) {
  if (!profile?.goals.trim()) return false;

  const subjectWords = meaningfulWords(
    `${subject.title} ${subject.details} ${subject.category}`,
  );
  const goalWords = meaningfulWords(profile.goals);

  for (const word of subjectWords) {
    if (goalWords.has(word)) return true;
  }

  return false;
}

function deriveSignals(
  subject: CompletionSubject,
  history: EncouragementRecord[],
  profile: UserProfile | null,
  completedAt: string,
): EncouragementSignal[] {
  const signals: EncouragementSignal[] = [];
  const completedBefore = history.length;

  if (subject.priority === 'urgent' || subject.priority === 'high') {
    signals.push('highPriority');
  }

  if (hoursBetween(subject.createdAt, completedAt) >= 72) {
    signals.push('delayed');
  }

  if (!history.some((record) => isSameLocalDay(record.createdAt, completedAt))) {
    signals.push('firstToday');
  }

  if ((completedBefore + 1) % 5 === 0) {
    signals.push('milestone');
  }

  const latest = history[0];
  if (latest && hoursBetween(latest.createdAt, completedAt) >= 48) {
    signals.push('returning');
  }

  if (isGoalAligned(subject, profile)) {
    signals.push('goalAligned');
  }

  if (subject.sourceType === 'focus') {
    signals.push('focused');
  }

  return signals;
}

function hashString(value: string) {
  let hash = 2166136261;

  for (let index = 0; index < value.length; index += 1) {
    hash ^= value.charCodeAt(index);
    hash = Math.imul(hash, 16777619);
  }

  return Math.abs(hash >>> 0);
}

function scorePrinciple(
  principle: EncouragementPrinciple,
  context: EncouragementContext,
  signals: EncouragementSignal[],
  history: EncouragementRecord[],
  preferences: EncouragementPreferences,
) {
  let score = 0;

  if (principle.contexts.includes(context)) score += 7;
  if (principle.contexts.includes('general')) score += 1;

  signals.forEach((signal) => {
    if (principle.signals.includes(signal)) score += 4;
  });

  const savedForPrinciple = history.filter(
    (record) => record.saved && record.principleId === principle.id,
  ).length;
  const dismissedForPrinciple = history.filter(
    (record) => record.dismissed && record.principleId === principle.id,
  ).length;

  score += Math.min(savedForPrinciple, 3) * 1.5;
  score -= Math.min(dismissedForPrinciple, 3);

  if (preferences.pinnedPrincipleIds.includes(principle.id)) {
    score += 3;
  }

  return score;
}

function namePrefix(
  profile: UserProfile | null,
  preferences: EncouragementPreferences,
) {
  const name = profile?.preferredName.trim();

  if (
    !preferences.usePreferredName ||
    !name ||
    profile?.isGuest ||
    name.toLowerCase() === 'guest'
  ) {
    return '';
  }

  return `${name}, `;
}

function composeMessage(
  subject: CompletionSubject,
  principle: EncouragementPrinciple,
  context: EncouragementContext,
  leadIndex: number,
  length: EncouragementLength,
  profile: UserProfile | null,
  preferences: EncouragementPreferences,
  goalAligned: boolean,
) {
  const definition = contextById.get(context) ?? contextById.get('general')!;
  const lead = definition.leads[leadIndex % definition.leads.length];
  const prefix = namePrefix(profile, preferences);
  const namedLead = prefix
    ? `${prefix}${lead.charAt(0).toLowerCase()}${lead.slice(1)}`
    : lead;

  const parts = [namedLead, principle.tones[preferences.tone]];

  if (length === 'deeper') {
    parts.push(principle.summary);

    if (goalAligned) {
      parts.push(
        'This completion also supports a direction you named for yourself.',
      );
    }
  }

  return parts.join(' ');
}

export function generateEncouragement({
  subject,
  profile,
  history,
  preferences,
  completedAt,
  recentMessageIds,
  excludedMessageIds = [],
}: {
  subject: CompletionSubject;
  profile: UserProfile | null;
  history: EncouragementRecord[];
  preferences: EncouragementPreferences;
  completedAt: string;
  recentMessageIds: string[];
  excludedMessageIds?: string[];
}): EncouragementDraft {
  const context = classifyContext(subject, profile);
  const signals = deriveSignals(subject, history, profile, completedAt);
  const goalAligned = signals.includes('goalAligned');
  const allowedPrinciples = encouragementPrinciples.filter(
    (principle) => preferences.spiritualEnabled || !principle.spiritual,
  );

  const scored = allowedPrinciples
    .map((principle) => ({
      principle,
      score: scorePrinciple(
        principle,
        context,
        signals,
        history,
        preferences,
      ),
    }))
    .sort((left, right) => right.score - left.score);

  const topCandidates = scored.slice(0, Math.min(8, scored.length));
  const seed = hashString(
    `${subject.sourceId}:${completedAt}:${preferences.tone}:${history.length}`,
  );
  const blocked = new Set([
    ...recentMessageIds.slice(0, 10),
    ...excludedMessageIds,
  ]);

  let selectedPrinciple = topCandidates[0]?.principle ?? allowedPrinciples[0];
  let selectedLeadIndex = seed % 3;
  let selectedMessageId = '';

  for (let attempt = 0; attempt < topCandidates.length * 3; attempt += 1) {
    const candidate =
      topCandidates[(seed + attempt) % topCandidates.length]?.principle ??
      selectedPrinciple;
    const definition =
      contextById.get(context) ?? contextById.get('general')!;
    const leadIndex = (seed + attempt) % definition.leads.length;
    const messageId = [
      candidate.id,
      context,
      preferences.tone,
      preferences.length,
      leadIndex,
    ].join(':');

    selectedPrinciple = candidate;
    selectedLeadIndex = leadIndex;
    selectedMessageId = messageId;

    if (!blocked.has(messageId)) break;
  }

  const message = composeMessage(
    subject,
    selectedPrinciple,
    context,
    selectedLeadIndex,
    preferences.length,
    profile,
    preferences,
    goalAligned,
  );

  return {
    context,
    principleId: selectedPrinciple.id,
    principleTitle: selectedPrinciple.title,
    messageId: selectedMessageId,
    heading:
      subject.sourceType === 'focus'
        ? 'Focus block complete'
        : `Completed: ${subject.title}`,
    message,
  };
}

export function shouldSurfaceEncouragement({
  subject,
  preferences,
  history,
  completedAt,
}: {
  subject: CompletionSubject;
  preferences: EncouragementPreferences;
  history: EncouragementRecord[];
  completedAt: string;
}) {
  switch (preferences.frequency) {
    case 'every':
      return true;
    case 'important':
      return (
        subject.priority === 'urgent' ||
        subject.priority === 'high' ||
        hoursBetween(subject.createdAt, completedAt) >= 72
      );
    case 'firstDaily':
      return !history.some((record) =>
        isSameLocalDay(record.createdAt, completedAt),
      );
    case 'milestones':
      return (history.length + 1) % 5 === 0;
    case 'off':
      return false;
    default:
      return true;
  }
}

export function getLessonsEarned(
  history: EncouragementRecord[],
): LessonSummary[] {
  const counts = new Map<
    string,
    { count: number; savedCount: number }
  >();

  history.forEach((record) => {
    const current = counts.get(record.principleId) ?? {
      count: 0,
      savedCount: 0,
    };

    current.count += 1;
    if (record.saved) current.savedCount += 1;
    counts.set(record.principleId, current);
  });

  return [...counts.entries()]
    .map(([principleId, countData]) => {
      const principle = principleById.get(principleId);
      if (!principle) return null;

      return {
        principleId,
        title: principle.title,
        summary: principle.summary,
        ...countData,
      };
    })
    .filter((value): value is LessonSummary => value !== null)
    .sort((left, right) => {
      if (right.savedCount !== left.savedCount) {
        return right.savedCount - left.savedCount;
      }

      return right.count - left.count;
    });
}

export function getAvailablePrinciples(spiritualEnabled: boolean) {
  return encouragementPrinciples.filter(
    (principle) => spiritualEnabled || !principle.spiritual,
  );
}

export function buildWeeklyReflection(
  data: AppData,
  now = new Date(),
): WeeklyReflection {
  const cutoff = new Date(now);
  cutoff.setDate(cutoff.getDate() - 6);
  cutoff.setHours(0, 0, 0, 0);

  const recentRecords = data.encouragementHistory.filter(
    (record) => new Date(record.createdAt) >= cutoff,
  );
  const recentTasks = data.tasks.filter(
    (task) =>
      task.completedAt !== null && new Date(task.completedAt) >= cutoff,
  );
  const recentFocus = data.focusSessions.filter(
    (session) => new Date(session.startedAt) >= cutoff,
  );

  const completedCount = recentTasks.length;
  const highPriorityCount = recentTasks.filter(
    (task) => task.priority === 'urgent' || task.priority === 'high',
  ).length;
  const focusCount = recentFocus.length;

  const contextCounts = new Map<EncouragementContext, number>();
  recentRecords.forEach((record) => {
    contextCounts.set(
      record.context,
      (contextCounts.get(record.context) ?? 0) + 1,
    );
  });

  const dominantContext = [...contextCounts.entries()].sort(
    (left, right) => right[1] - left[1],
  )[0]?.[0];
  const dominantContextLabel = dominantContext
    ? contextById.get(dominantContext)?.label ?? null
    : null;

  if (completedCount === 0 && focusCount === 0) {
    return {
      heading: 'The week is still open',
      message:
        'No judgment is required here. Choose one honest next step, make it small enough to begin, and let action create the first evidence of movement.',
      completedCount,
      highPriorityCount,
      focusCount,
      dominantContextLabel,
    };
  }

  const parts = [
    `You completed ${completedCount} ${
      completedCount === 1 ? 'task' : 'tasks'
    } during the last seven days.`,
  ];

  if (highPriorityCount > 0) {
    parts.push(
      `${highPriorityCount} ${
        highPriorityCount === 1
          ? 'was a high-priority commitment'
          : 'were high-priority commitments'
      }.`,
    );
  }

  if (focusCount > 0) {
    parts.push(
      `You also protected ${focusCount} ${
        focusCount === 1 ? 'focus block' : 'focus blocks'
      }.`,
    );
  }

  if (dominantContextLabel) {
    parts.push(
      `Your strongest pattern of movement was in ${dominantContextLabel.toLowerCase()}.`,
    );
  }

  const returned = recentRecords.some(
    (record) => record.principleId === 'returning-beats-perfection',
  );

  if (returned) {
    parts.push(
      'The important pattern was not perfection; it was your willingness to return.',
    );
  } else {
    parts.push(
      'Progress did not need to be dramatic to remain real and useful.',
    );
  }

  return {
    heading: 'Evidence from your week',
    message: parts.join(' '),
    completedCount,
    highPriorityCount,
    focusCount,
    dominantContextLabel,
  };
}

File: ECHIRON-Android/app/src/main/java/com/flameborne/echiron/model/EncouragementEngine.kt
package com.flameborne.echiron.model

import java.time.Instant
import java.time.ZoneId
import kotlin.math.exp
import kotlin.math.max
import kotlin.math.min

data class CompletionSubject(
    val sourceId: String,
    val title: String,
    val details: String,
    val category: String,
    val priority: Priority,
    val createdAt: String,
    val focusSession: Boolean = false,
)

data class EncouragementDraft(
    val context: EncouragementContext,
    val secondaryContext: EncouragementContext?,
    val principleId: String,
    val principleTitle: String,
    val principleSource: String,
    val principleSourceId: String,
    val matchedSignals: Set<EncouragementSignal>,
    val messageId: String,
    val heading: String,
    val message: String,
)

private data class ContextMatch(
    val primary: EncouragementContext,
    val secondary: EncouragementContext?,
    val scores: Map<EncouragementContext, Double>,
)

private data class ScoredPrinciple(
    val principle: EncouragementPrinciple,
    val score: Double,
)

private data class Composition(
    val message: String,
    val messageId: String,
)

object EncouragementEngine {
    private val stopWords =
        setOf(
            "about", "after", "again", "also", "and", "are", "because", "been", "being",
            "but", "can", "could", "did", "does", "for", "from", "had", "has", "have",
            "into", "its", "more", "not", "only", "that", "the", "their", "them", "then",
            "there", "these", "they", "this", "through", "too", "was", "were", "what",
            "when", "where", "which", "while", "will", "with", "would", "your",
        )
    private val uncertainKeywords =
        listOf("afraid", "anxious", "avoid", "conflict", "decide", "difficult", "doubt", "fear", "finally", "hard", "maybe", "nervous", "stuck", "unsure", "worry")
    private val restorativeKeywords =
        listOf("break", "doctor", "exercise", "heal", "health", "meal", "medication", "pause", "recover", "rest", "sleep", "stretch", "walk", "water", "workout")
    private val foundationKeywords =
        listOf("budget", "calendar", "clean", "foundation", "maintenance", "organize", "plan", "prepare", "schedule", "setup", "system", "tax")
    private val sharedKeywords =
        listOf("care", "child", "client", "community", "family", "friend", "help", "neighbor", "partner", "support", "team", "together", "volunteer")
    private val practiceKeywords =
        listOf("again", "daily", "habit", "practice", "repeat", "routine", "session", "weekly")

    fun forCompletedTask(
        task: EchironTask,
        profile: UserProfile,
        history: List<Encouragement>,
        preferences: EncouragementPreferences,
        completedAt: String,
        excludedMessageIds: Set<String> = emptySet(),
    ): EncouragementDraft =
        generate(
            subject =
                CompletionSubject(
                    sourceId = task.id,
                    title = task.title,
                    details = task.details,
                    category = task.category,
                    priority = task.priority,
                    createdAt = task.createdAt,
                ),
            profile = profile,
            history = history,
            preferences = preferences,
            completedAt = completedAt,
            excludedMessageIds = excludedMessageIds,
        )

    fun forFocusSession(
        sourceId: String,
        minutes: Int,
        profile: UserProfile,
        history: List<Encouragement>,
        preferences: EncouragementPreferences,
        completedAt: String,
        excludedMessageIds: Set<String> = emptySet(),
    ): EncouragementDraft =
        generate(
            subject =
                CompletionSubject(
                    sourceId = sourceId,
                    title = "$minutes-minute focus block",
                    details = "A protected block of focused, distraction-limited work.",
                    category = "Focus",
                    priority = Priority.MEDIUM,
                    createdAt = completedAt,
                    focusSession = true,
                ),
            profile = profile,
            history = history,
            preferences = preferences,
            completedAt = completedAt,
            excludedMessageIds = excludedMessageIds,
        )

    fun generate(
        subject: CompletionSubject,
        profile: UserProfile,
        history: List<Encouragement>,
        preferences: EncouragementPreferences,
        completedAt: String,
        excludedMessageIds: Set<String> = emptySet(),
    ): EncouragementDraft {
        val contexts = classifyContexts(subject, profile)
        val signals = deriveSignals(subject, contexts, history, profile, completedAt)
        val subjectText =
            normalize("${subject.title} ${subject.details} ${subject.category} ${profile.goals} ${profile.role}")
        val allowed = EncouragementCatalog.principles.filter { preferences.spiritualEnabled || !it.spiritual }
        require(allowed.isNotEmpty()) { "The encouragement catalog must contain an enabled principle." }
        val seed =
            hashString(
                "${subject.sourceId}:$completedAt:${preferences.tone}:${preferences.length}:${history.size}:${excludedMessageIds.sorted().joinToString("|")}",
            )
        val scored =
            allowed
                .map {
                    ScoredPrinciple(
                        it,
                        scorePrinciple(it, contexts, signals, subjectText, history, preferences, seed),
                    )
                }
                .sortedByDescending { it.score }
                .take(40)
        val maximum = scored.first().score
        val relevant = scored.filter { it.score >= maximum - 9.0 }
        val weighted = relevant.map { it to exp((it.score - maximum) / 3.4) }
        var target = seededUnit(seed xor 0x05ebca6b) * weighted.sumOf { it.second }
        var selected = weighted.first().first
        for ((candidate, weight) in weighted) {
            target -= weight
            if (target <= 0.0) {
                selected = candidate
                break
            }
        }

        val blocked =
            (history.take(40).map { it.messageId } + excludedMessageIds).toSet()
        val recentMessages = history.take(24).map { it.message }
        var best: Pair<Composition, Double>? = null
        repeat(72) { attempt ->
            val composition =
                compose(
                    selected.principle,
                    contexts,
                    signals,
                    profile,
                    preferences,
                    seed,
                    attempt,
                )
            val similarity = recentMessages.maxOfOrNull { jaccard(composition.message, it) } ?: 0.0
            val utility = -similarity * 11.0 - if (composition.messageId in blocked) 100.0 else 0.0
            if (utility > (best?.second ?: Double.NEGATIVE_INFINITY)) best = composition to utility
        }
        val composition = checkNotNull(best).first
        return EncouragementDraft(
            context = contexts.primary,
            secondaryContext = contexts.secondary,
            principleId = selected.principle.id,
            principleTitle = selected.principle.title,
            principleSource = selected.principle.sourceLabel,
            principleSourceId = selected.principle.sourceId,
            matchedSignals = signals,
            messageId = composition.messageId,
            heading = if (subject.focusSession) "Focus block complete" else "Completed: ${subject.title}",
            message = composition.message,
        )
    }

    private fun classifyContexts(subject: CompletionSubject, profile: UserProfile): ContextMatch {
        val category = normalize(subject.category)
        val title = normalize(subject.title)
        val details = normalize(subject.details)
        val profileText = normalize("${profile.role} ${profile.goals}")
        val scores =
            EncouragementCatalog.contexts
                .filterKeys { it != EncouragementContext.GENERAL }
                .mapValues { (_, definition) ->
                    var score = if (category == normalize(definition.label)) 8.0 else 0.0
                    definition.keywords.forEach { keyword ->
                        if (includesKeyword(category, keyword)) score += 5.0
                        if (includesKeyword(title, keyword)) score += 3.0
                        if (includesKeyword(details, keyword)) score += 1.5
                        if (includesKeyword(profileText, keyword)) score += 0.4
                    }
                    score
                }.toMutableMap()
        if (subject.focusSession) scores[EncouragementContext.FOCUS] = (scores[EncouragementContext.FOCUS] ?: 0.0) + 100.0
        val ranked = scores.filterValues { it > 0.0 }.entries.sortedByDescending { it.value }
        val primary = ranked.firstOrNull()?.key ?: EncouragementContext.GENERAL
        val firstScore = ranked.firstOrNull()?.value ?: 0.0
        val secondary =
            ranked.firstOrNull {
                it.key != primary && it.value >= max(3.0, firstScore * 0.22)
            }?.key
        return ContextMatch(primary, secondary, scores)
    }

    private fun deriveSignals(
        subject: CompletionSubject,
        contexts: ContextMatch,
        history: List<Encouragement>,
        profile: UserProfile,
        completedAt: String,
    ): Set<EncouragementSignal> =
        buildSet {
            val text = normalize("${subject.title} ${subject.details} ${subject.category}")
            val age = hoursBetween(subject.createdAt, completedAt)
            if (subject.priority == Priority.URGENT || subject.priority == Priority.HIGH) add(EncouragementSignal.HIGH_PRIORITY)
            if (age >= 72) add(EncouragementSignal.DELAYED)
            if (age >= 168) add(EncouragementSignal.LONG_CARRIED)
            if (history.none { sameLocalDay(it.createdAt, completedAt) }) add(EncouragementSignal.FIRST_TODAY)
            if ((history.size + 1) % 5 == 0) add(EncouragementSignal.MILESTONE)
            if (history.firstOrNull()?.let { hoursBetween(it.createdAt, completedAt) >= 48 } == true) add(EncouragementSignal.RETURNING)
            if (profile.goals.isNotBlank() && overlaps(text, profile.goals)) add(EncouragementSignal.GOAL_ALIGNED)
            if (subject.focusSession || contexts.primary == EncouragementContext.FOCUS) add(EncouragementSignal.FOCUSED)
            if (containsAny(text, uncertainKeywords) || contexts.primary in setOf(EncouragementContext.COURAGE, EncouragementContext.TRANSITION)) add(EncouragementSignal.UNCERTAIN_PATH)
            if (containsAny(text, restorativeKeywords) || contexts.primary == EncouragementContext.RECOVERY) add(EncouragementSignal.RESTORATIVE)
            if (containsAny(text, foundationKeywords) || contexts.primary in setOf(EncouragementContext.PLANNING, EncouragementContext.HOME, EncouragementContext.FINANCE)) add(EncouragementSignal.FOUNDATION_BUILDING)
            if (containsAny(text, sharedKeywords) || contexts.primary in setOf(EncouragementContext.FAMILY, EncouragementContext.COMMUNITY, EncouragementContext.CAREGIVING)) add(EncouragementSignal.SHARED_BENEFIT)
            if (contexts.primary == EncouragementContext.CREATIVE || contexts.secondary == EncouragementContext.CREATIVE) add(EncouragementSignal.CREATIVE_OUTPUT)
            if (contexts.primary == EncouragementContext.EDUCATION || contexts.secondary == EncouragementContext.EDUCATION) add(EncouragementSignal.LEARNING_PROGRESS)
            if (EncouragementSignal.GOAL_ALIGNED in this || contexts.primary == EncouragementContext.GROWTH) add(EncouragementSignal.IDENTITY_ALIGNED)
            val related = history.take(40).count { it.context == contexts.primary || overlaps(it.subjectTitle, subject.title) }
            if (related >= 2 || containsAny(text, practiceKeywords)) add(EncouragementSignal.STEADY_PRACTICE)
        }

    private fun scorePrinciple(
        principle: EncouragementPrinciple,
        contexts: ContextMatch,
        signals: Set<EncouragementSignal>,
        subjectText: String,
        history: List<Encouragement>,
        preferences: EncouragementPreferences,
        seed: Int,
    ): Double {
        var score = 0.0
        if (contexts.primary in principle.contexts) score += 8.0
        if (contexts.secondary != null && contexts.secondary in principle.contexts) score += 3.5
        if (EncouragementContext.GENERAL in principle.contexts) score += 1.0
        score += signals.count { it in principle.signals } * 3.25
        score += min(5, principle.keywords.count { includesKeyword(subjectText, it) }) * 1.4
        val savedPrinciple = history.count { it.saved && it.principleId == principle.id }
        val dismissedPrinciple = history.count { it.dismissed && it.principleId == principle.id }
        val savedSource = history.count { it.saved && it.principleSourceId == principle.sourceId }
        val dismissedSource = history.count { it.dismissed && it.principleSourceId == principle.sourceId }
        score += min(savedPrinciple, 4) * 2.0
        score -= min(dismissedPrinciple, 4) * 2.5
        score += min(savedSource, 6) * 0.55
        score -= min(dismissedSource, 6) * 0.65
        if (principle.id in preferences.pinnedPrincipleIds) score += 5.0
        val recent = history.take(18)
        val recentIndex = recent.indexOfFirst { it.principleId == principle.id }
        if (recentIndex >= 0) {
            val repetition = max(1.5, 7.0 - recentIndex * 0.45)
            val factor = if (principle.id in preferences.pinnedPrincipleIds) 0.1 else if (savedPrinciple > 0) 0.3 else 1.0
            score -= repetition * factor
        }
        if (history.none { it.principleId == principle.id }) score += 0.8
        score -= min(recent.count { it.principleSourceId == principle.sourceId }, 5) * 0.35
        score += seededUnit(seed xor hashString(principle.id)) * 1.2
        return score
    }

    private fun compose(
        principle: EncouragementPrinciple,
        contexts: ContextMatch,
        signals: Set<EncouragementSignal>,
        profile: UserProfile,
        preferences: EncouragementPreferences,
        seed: Int,
        attempt: Int,
    ): Composition {
        val definition = checkNotNull(EncouragementCatalog.contexts[contexts.primary])
        val selectedSignal = signals.firstOrNull { it in principle.signals } ?: signals.firstOrNull()
        val evidence = selectedSignal?.let { EncouragementCatalog.signalEvidence[it] }.orEmpty()
        val closings = checkNotNull(EncouragementCatalog.toneClosings[preferences.tone])
        val plan = checkNotNull(EncouragementCatalog.planningClosings[profile.planningStyle])
        val leadIndex = index(seed + attempt * 7, definition.leads.size)
        val evidenceIndex = if (evidence.isEmpty()) -1 else index(seed + attempt * 13, evidence.size)
        val practiceIndex = if (principle.practices.isEmpty()) -1 else index(seed + attempt * 17, principle.practices.size)
        val closingIndex = index(seed + attempt * 19, closings.size)
        val planIndex = index(seed + attempt * 23, plan.size)
        val bridgeIndex = index(seed + attempt * 29, EncouragementCatalog.contextBridges.size)
        val nameIndex = index(seed + attempt * 31, 3)
        val lead = definition.leads[leadIndex]
        val name = profile.preferredName.trim().takeIf { preferences.usePreferredName && it.isNotBlank() }
        val opening =
            when {
                name == null -> lead
                nameIndex == 0 -> "$name, ${lead.replaceFirstChar { it.lowercase() }}"
                nameIndex == 1 -> "$lead Take that in, $name."
                else -> "$lead $name, this one deserves an honest acknowledgment."
            }
        val parts = mutableListOf(opening, checkNotNull(principle.tones[preferences.tone]))
        if (evidenceIndex >= 0) parts += evidence[evidenceIndex]
        if (preferences.length == EncouragementLength.DEEPER) {
            val summary = principle.summary.replaceFirstChar { it.lowercase() }
            parts += "${EncouragementCatalog.contextBridges[bridgeIndex]} $summary"
            if (practiceIndex >= 0) parts += "A useful way to carry this forward: ${principle.practices[practiceIndex]}"
            parts += plan[planIndex]
        } else {
            parts += closings[closingIndex]
        }
        val messageId =
            listOf(
                principle.id,
                contexts.primary,
                contexts.secondary ?: "none",
                preferences.tone,
                preferences.length,
                leadIndex,
                selectedSignal ?: "none",
                evidenceIndex,
                practiceIndex,
                if (preferences.length == EncouragementLength.DEEPER) planIndex else closingIndex,
                if (name == null) -1 else nameIndex,
            ).joinToString(":")
        return Composition(parts.joinToString(" "), messageId)
    }

    private fun normalize(value: String): String =
        value.lowercase().replace(Regex("[^a-z0-9\\s-]"), " ").replace(Regex("\\s+"), " ").trim()

    private fun meaningfulWords(value: String): Set<String> =
        normalize(value).split(Regex("\\s+")).filter { it.length >= 4 && it !in stopWords }.toSet()

    private fun includesKeyword(text: String, keyword: String): Boolean {
        val clean = normalize(keyword)
        if (clean.isBlank()) return false
        return if (' ' in clean || '-' in clean) clean in text else clean in text.split(' ').toSet()
    }

    private fun containsAny(text: String, keywords: List<String>): Boolean = keywords.any { includesKeyword(text, it) }

    private fun overlaps(left: String, right: String): Boolean =
        meaningfulWords(left).intersect(meaningfulWords(right)).isNotEmpty()

    private fun jaccard(left: String, right: String): Double {
        val leftWords = meaningfulWords(left)
        val rightWords = meaningfulWords(right)
        if (leftWords.isEmpty() || rightWords.isEmpty()) return 0.0
        return leftWords.intersect(rightWords).size.toDouble() / leftWords.union(rightWords).size
    }

    private fun hoursBetween(earlier: String, later: String): Long =
        runCatching {
            max(0L, java.time.Duration.between(Instant.parse(earlier), Instant.parse(later)).toHours())
        }.getOrDefault(0L)

    private fun sameLocalDay(left: String, right: String): Boolean =
        runCatching {
            val zone = ZoneId.systemDefault()
            Instant.parse(left).atZone(zone).toLocalDate() == Instant.parse(right).atZone(zone).toLocalDate()
        }.getOrDefault(false)

    private fun hashString(value: String): Int {
        var hash = 0x811c9dc5L
        value.forEach { char ->
            hash = hash xor char.code.toLong()
            hash = (hash * 16777619L) and 0xffffffffL
        }
        return hash.toInt()
    }

    private fun seededUnit(seed: Int): Double {
        val value = (1664525L * (seed.toLong() and 0xffffffffL) + 1013904223L) and 0xffffffffL
        return value.toDouble() / 4294967296.0
    }

    private fun index(value: Int, size: Int): Int = Math.floorMod(value, size)
}
